Rewrite logic for Deku Backroom GS to be human-readable and change the basement web trick

DuskTheUmbreon opened the pull request:
I rewrote some of the here()s for the to just be a room instead, and sorted it all out to be less of a mess. This should be the same logically, just far more readable and easier to understand.

~~I've drafted it because I'm not sure that the trick for burning the vertical webs is actually really that important. 2/3 fire requirements are absolutely trivial with bow (one is literally the exact same as the intended method for torches room in water temple, and the other is even easier), and the only slightly awkward one is in the main b1 room since you need to stand to the side of the torch.~~

I've now changed it to make the 'Deku Tree Basement Webs with Bow' trick just for Gohma, after consensus from r0b and Rattus that burning the two vertical webs in the basement with bow was easy enough to not be justified as a trick.

DuskTheUmbreon committed:
Rewrite Deku backroom GS to be more readable.

I rewrote some of the here()s to just be a room instead, and sorted it all out to be less of a mess.

This should be the same logically, just actually human-readable.

I've drafted it because I'm not sure that the trick for burning the vertical webs is actually necessary. 2/3 fire requirements are absolutely trivial with bow (one is literally the exact same as the intended method for torches room in water temple, and the other is even easier), and the only slightly awkward one is in the main b1 room (you need to stand to the right of the torch and shoot. It doesn't look like the fire will go through because you're to the side of the torch, but it does because OoT).

DuskTheUmbreon committed:
Update Deku Tree.json

I forgot how to whitespace properly

r0bd0g commented:
For some reason, idk why, regions in dungeons generally exit back out if they can.
The line for the exit to the backroom seems awfully long on one line...
You should probably reword the deku basement webs trick in settingslist.

DuskTheUmbreon committed:
Updates per r0b's suggestions

DuskTheUmbreon commented:
Added an exit per r0b's suggestion and split the entry to the room to more lines.

I'll hold off on changing settingslist until there's a definitive decision on if the trick should apply to the vertical webs.

r0bd0g commented:
I'd recommend tabbing in line 25 and untabbing line 48.

Rattus says he included both webs as part of the trick for consistency reasons.
I can't remember if I tested the webs to the back skull with bow while there was a big chest there. I definitely did test the webs to Gohma, though.

r0bd0g commented:
Taking a closer look at this, you are missing the here(is_adult) alternative to deku b1 skip. (This is where adult climbs up and pushes the block down for child.)

fenhl commented on data/World/Deku Tree.json:
```diff
@@ -46,6 +41,19 @@
         }
     },
+    {
+        "region_name": "Deku Tree Basement Backroom"
```
```suggestion
        "region_name": "Deku Tree Basement Backroom",
```

DuskTheUmbreon committed:
Add missing comma

Thanks, fenhl. I thought I had caught all of them, turns out I was wrong.

Co-authored-by: Fenhl <fenhl@fenhl.net>

DuskTheUmbreon committed:
Typo fixes

Removed a parenthesis on line 24, added the case for adult opening the way for child, and fixed a tab issue on line 48.

I thought I had caught them all before I put in the PR, unfortunately not.

DuskTheUmbreon commented:
Fixed the typos r0b mentioned, and added the missing logic for adult pushing the block for child, but I don't see a tabbing issue on line 25...It looks like line 29 has an extra tab, though? Or is that how the "(a or b) and c" multiline logic is supposed to look?

r0bd0g commented:
I dunno, generally I tab in if I open a parenthesis but haven't closed it.

Alright this is essentially what was there before:
```            "Deku Tree GS Basement Back Room": "
                (here(has_fire_source_with_torch or (can_use(Bow) and logic_deku_b1_webs_with_bow)) and
                    here(can_use(Slingshot) or can_use(Bow)) and
                    here(can_blast_or_smash) and
                    (can_use(Hookshot) or can_use(Boomerang))) or
                ((logic_deku_b1_skip or here(is_adult)) and is_child and
                    has_explosives and can_use(Boomerang) and
                    (Sticks or can_use(Dins_Fire)))"
```
And what you changed it to is:
```            "Deku Tree GS Basement Back Room": "
                ((here(has_fire_source_with_torch or (can_use(Bow) and logic_deku_b1_webs_with_bow)) and
                        here(can_use(Slingshot) or can_use(Bow))) or
                    ((logic_deku_b1_skip or here(is_adult)) and is_child)) and
                here(has_fire_source_with_torch or can_use(Bow)) and
                here(can_blast_or_smash) and
                (can_use(Boomerang) or can_use(Hookshot))"
```
But something to notice there is that fire_source/bow is being double checked. The only difference is the trick only being checked the first time. No matter how you burn that first web, you can use the same method on the following webs. The exception is if you took the crawlspace, but then adult can't get back there to use the bow, which make whether you check the trick there irrelevant. So you could optimize a bit with:
```
            "Deku Tree GS Basement Back Room": "
                (here(can_use(Slingshot) or can_use(Bow)) or
                    ((logic_deku_b1_skip or here(is_adult)) and is_child)) and
                here(has_fire_source_with_torch or (can_use(Bow) and logic_deku_b1_webs_with_bow)) and
                here(can_blast_or_smash) and
                (can_use(Boomerang) or can_use(Hookshot))"
```
If you must implement this with regions you could somewhat-confusingly only check that full firesourcetorch/bow+trick inside the back room region. So it'll hurt your readability a bit.

r0bd0g commented:
I dunno, generally I tab in if I open a parenthesis but haven't closed it.

I think you could optimize this a bit by not checking firesource to get to the backroom. Just put the full here:firesourcetorch/bow+trick logic on the skull directly. It's kind of a heavy section of code to be repeating? But doing that might hurt your readability a bit.

DuskTheUmbreon commented:
On the tab: Makes sense, I can change that if it helps readability, but I've been going with only indenting once for that kind of situation personally.

I could optimize it a bit by removing the extra here(), but I don't feel the slight optimization is worth the loss of readability, since the entire purpose of this was to make it more readable.

r0bd0g commented:
Let's go with default logic for using bow to get to the backroom. Remember to update settingslist as needed as well.

DuskTheUmbreon committed:
Make burning the vertical b1 web default logic

DuskTheUmbreon committed:
Change the relevant trick for webs

Since it's now default logic, having the trick reference it would be a bad idea.

DuskTheUmbreon committed:
Update the changelog to reflect the logic change

I'm not 100% sure if this is the best way to phrase this, so someone may want to reword it.

DuskTheUmbreon commented:
Alright. Updated the logic, and the trick itself in settingslist.py (I'm not 100% happy with my wording but it's at least correct) and noted it in the the changelog (Again, not 100% happy with the wording, but at least it's accurate).

Roman971 commented:
I suggest renaming the PR and adding a note to the description so this doesn't get mistaken for a simple rewrite.

DuskTheUmbreon committed:
Phrasing change for the trick

Thanks ETR.

DuskTheUmbreon commented:
Changed the wording for the trick (thanks ETR) and updated the trick name per Roman's suggestion.

Zannick commented:
Looks good mechanically; I'll fix the tabbing to my liking. Thanks for adding the readme entry, too!

Is it better to say "circular web" than "floor web" to clarify what web we're talking about? Just a minor thing.
